### PR TITLE
Model program

### PR DESCRIPTION
**Pull request by sblack-usu**

### CODE REVIEW

File: hsmodels/schemas/aggregations.py [hunk 1]
```diff
@@ -296,3 +303,66 @@ class Config:
...
     _parse_abstract = root_validator(pre=True, allow_reuse=True)(parse_abstract)
+
+
+class ModelProgramMetadata(BaseAggregationMetadata):
+    """
+    A class used to represent the metadata associated with a model program aggregation
+    """
+
+    class Config:
+        title = 'Single File Aggregation Metadata'
```
Comment (horsburgh) [inline hsmodels/schemas/aggregations.py:314]: I'm assuming this should be "Model Program Aggregation Metadata"

File: hsmodels/schemas/aggregations.py [hunk 2]
```diff
@@ -296,3 +303,66 @@ class Config:
...
+        title="Aggregation type",
+        description="A string expressing the aggregation type from the list of HydroShare aggregation types",
+        allow_mutation=False,
+    )
+
+    version: str = Field(
+        default=None, title="Version", description="The software version or build number of the model", max_length=255
+    )
+
+    name: str = Field(default="Unknown Model Program", max_length=255, title="Model Program Name", description="TODO")
```
Comment (horsburgh) [inline hsmodels/schemas/aggregations.py:330]: Change description to "A string expressing the name of the model program"

File: hsmodels/schemas/aggregations.py [hunk 3]
```diff
@@ -296,3 +303,66 @@ class Config:
...
+        title="Operating Systems",
+        description="Compatible operating systems to setup and run the model",
+    )
+
+    release_date: date = Field(
+        default=None, title="Release Date", description="The date that this version of the model was released"
+    )
+
+    website: AnyUrl = Field(
+        default=None, title='Website', description='A URL to the website maintained by the model developers'
```
Comment (horsburgh) [inline hsmodels/schemas/aggregations.py:351]: Change the description to "A URL to a website describing the model that is maintained by the model developers"

File: hsmodels/schemas/aggregations.py [hunk 4]
```diff
@@ -296,3 +303,66 @@ class Config:
...
+    )
+
+    code_repository: AnyUrl = Field(
+        default=None,
+        title='Software Repository',
+        description='A URL to the source code repository (e.g. git, mercurial, svn)',
```
Comment (horsburgh) [inline hsmodels/schemas/aggregations.py:357]: Change the description to "A URL to the source code repository for the model code (e.g., git, mercurial, svn, etc.)"

File: hsmodels/schemas/aggregations.py [hunk 5]
```diff
@@ -296,3 +303,66 @@ class Config:
...
+    )
+
+    file_types: List[ModelProgramFile] = Field(
+        default=None, title='File Types', description='A URL to the source code repository (e.g. git, mercurial, svn)'
```
Comment (horsburgh) [inline hsmodels/schemas/aggregations.py:361]: I'm assuming this description should be something like "File types used by the model program"

File: hsmodels/schemas/aggregations.py [hunk 6]
```diff
@@ -296,3 +303,66 @@ class Config:
...
+    )
+
+    json_schema: AnyUrl = Field(
+        default=None, title='Model program schema', description='A path to the model program JSON schema'
```
Comment (horsburgh) [inline hsmodels/schemas/aggregations.py:365]: Would it be more correct to say in the description "A path to the JSON metadata schema for the model program"

File: hsmodels/schemas/fields.py [hunk 7]
```diff
@@ -948,3 +948,15 @@ def start_before_end(cls, values):
         elif end and not start:
             raise ValueError(f"A start date was not included with end date [{end}]")
         return values
+
+
+class ModelProgramFile(BaseMetadata):
+    """"""
```
Comment (horsburgh) [inline hsmodels/schemas/fields.py:954]: """
    A class used to represent the metadata associated with a file used by a model program aggregation
"""

File: hsmodels/schemas/fields.py [hunk 8]
```diff
@@ -948,3 +948,15 @@ def start_before_end(cls, values):
...
+
+    class Config:
+        title = ""
```
Comment (horsburgh) [inline hsmodels/schemas/fields.py:957]: title = 'Model program file metadata'

File: hsmodels/schemas/fields.py [hunk 9]
```diff
@@ -948,3 +948,15 @@ def start_before_end(cls, values):
...
+
+    type: ModelProgramFileType = Field(
+        title="Model program file type", description="The type of file in a model program"
```
Comment (horsburgh) [inline hsmodels/schemas/fields.py:960]: I suggest description = 'The type of the file used by the model program'

File: hsmodels/schemas/fields.py [hunk 10]
```diff
@@ -948,3 +948,15 @@ def start_before_end(cls, values):
...
+    )
+    value: str = Field(title="Model program file name", description="The file name of a file with a labeled type")
```
Comment (horsburgh) [inline hsmodels/schemas/fields.py:962]: I suggest description = 'The name of the file used by the model program'

File: hsmodels/schemas/aggregations.py [hunk 11]
```diff
@@ -296,3 +303,109 @@ class Config:
...
+
+    includes_model_output: bool = Field(
+        title="Includes Model Output",
+        description="Indicates whether model output files are included in the aggregation",
+    )
+
+    executed_by: AnyUrl = Field(default=None, title="Executed By", description="A url to the Model Program")
+
+    program_schema_json: AnyUrl = Field(
+        default=None, title="Schema JSON url", description="A url to the Model Program schema json"
```
Comment (horsburgh) [inline hsmodels/schemas/aggregations.py:406]: I suggest:
title = "JSON Metadata schema URL"
description = "A URL to the JSON metadata schema for the related model program"

File: hsmodels/schemas/aggregations.py [hunk 12]
```diff
@@ -296,3 +303,109 @@ class Config:
...
+    )
+
+    program_schema_json_values: AnyUrl = Field(
+        default=None, title="Schema JSON values url", description="A url to the Model Program schema values json"
```
Comment (horsburgh) [inline hsmodels/schemas/aggregations.py:410]: I suggest:
title = "JSON metadata schema values URL"
description = "A URL to a JSON file containing the metadata values conforming to the JSON metadata schema for the related model program"

File: tests/data/metadata/modelinstance_meta.xml [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+      <dcterms:point>
+        <rdf:value>name=asdf; projection=WGS 84 EPSG:4326; east=-98.1969; north=41.85; units=Decimal degrees</rdf:value>
+      </dcterms:point>
+    </dc:coverage>
+    <dc:coverage>
+      <dcterms:period>
+        <rdf:value>start=2021-10-08T00:00:00; end=2021-10-15T00:00:00</rdf:value>
+      </dcterms:period>
+    </dc:coverage>
+    <hsterms:executedByModelProgram rdf:resource="https://www.hydroshare.org/resource/636729f24c03497e9b844ec3be63e823/data/contents/generic_file_resmap.xml#aggregation"/>
```
Comment (horsburgh) [inline tests/data/metadata/modelinstance_meta.xml:45]: This triple asserts that the model instance is executed by itself. Shouldn't this point at a model program aggregation?

File: tests/data/metadata/modelprogram_meta.xml [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  xmlns:dc="http://purl.org/dc/elements/1.1/"
+  xmlns:hsterms="https://www.hydroshare.org/terms/"
+  xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
+  xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
+>
+  <hsterms:ModelProgramAggregation rdf:about="https://www.hydroshare.org/resource/526e62aac1044285abc2cae4e2f463da/data/contents/modelprogram_resmap.xml#aggregation">
+    <hsterms:modelOperatingSystem>linux</hsterms:modelOperatingSystem>
+    <dc:title>setup againas</dc:title>
+    <hsterms:modelWebsite rdf:resource="https://www.hydroshare.org"/>
+    <hsterms:modelReleaseNotes rdf:resource="https://www.hydroshare.org/resource/526e62aac1044285abc2cae4e2f463da/data/contents/setup.cfg"/>
```
Comment (horsburgh) [inline tests/data/metadata/modelprogram_meta.xml:12]: Should modelReleaseNotes, modelDocumentation, modelSoftware, and modelEngine be repeatable?

**Review by horsburgh — APPROVED** [review]
Other than the comment about the repeatability of modelReleaseNotes, modelDocumentation, modelSoftware, and modelEngine in the model program RDF/XML example, the changes look good.

**Comment by pkdash** [comment]
@sblack-usu Didn't get time this week to review this. Can do on Monday if you want to wait.

File: tests/test_schemas.py [hunk 15]
```diff
@@ -7,6 +7,7 @@
     FileSetMetadata,
     GeographicFeatureMetadata,
     GeographicRasterMetadata,
+    ModelProgramMetadata,
```
Comment (pkdash) [inline tests/test_schemas.py:10]: Not seeing `ModelInstanceMetadata` as part of the schema test.

**Review by pkdash — COMMENTED** [review]
Left one comment. Otherwise looks great!

**Pull request merged**